File: src/main/java/com/teamwizardry/librarianlib/api/gui/components/template/ButtonTemplate.java
package com.teamwizardry.librarianlib.api.gui.components.template;

import com.teamwizardry.librarianlib.api.gui.GuiStyle;
import com.teamwizardry.librarianlib.api.gui.components.ComponentCenterAlign;
import com.teamwizardry.librarianlib.api.gui.components.ComponentSpriteTiled;
import com.teamwizardry.librarianlib.api.gui.components.ComponentVoid;
import com.teamwizardry.librarianlib.math.BoundingBox2D;
import com.teamwizardry.librarianlib.math.Vec2;

public class ButtonTemplate extends ComponentTemplate<ComponentVoid> {
	
	protected GuiStyle style;
	protected ComponentSpriteTiled tiled;
	public ComponentVoid contents;
	
	public ButtonTemplate(GuiStyle style, int posX, int posY) {
		
		result = new ComponentVoid(posX, posY);
		contents = new ComponentVoid(style.BUTTON_BORDER, style.BUTTON_BORDER);
		
		tiled = new ComponentSpriteTiled(style.BUTTON, style.BUTTON_BORDER, 0, 0);
		tiled.setSize(new Vec2(0, 0));
		
		result.add(tiled);
		result.add(contents);
		
		tiled.preDraw.add((c, pos, ticks) -> {
			BoundingBox2D box = contents.getLogicalSize();
			tiled.setSize(new Vec2(( box.width() - 1 ) + ( 2 * style.BUTTON_BORDER ), ( box.height() - 1 ) + ( 2 * style.BUTTON_BORDER )));
		});
	}
	
}

File: src/main/java/com/teamwizardry/librarianlib/client/gui/template/ButtonTemplate.kt
package com.teamwizardry.librarianlib.client.gui.template

import com.teamwizardry.librarianlib.client.gui.GuiComponent
import com.teamwizardry.librarianlib.client.gui.GuiStyle
import com.teamwizardry.librarianlib.client.gui.components.ComponentSpriteTiled
import com.teamwizardry.librarianlib.client.gui.components.ComponentVoid
import com.teamwizardry.librarianlib.common.util.vec

class ButtonTemplate(style: GuiStyle, posX: Int, posY: Int) : ComponentTemplate<ComponentVoid>(ComponentVoid(posX, posY)) {

    protected var style: GuiStyle? = null
    protected var tiled: ComponentSpriteTiled
    var contents: ComponentVoid

    init {

        contents = ComponentVoid(style.BUTTON_BORDER, style.BUTTON_BORDER)

        tiled = ComponentSpriteTiled(style.BUTTON, style.BUTTON_BORDER, 0, 0)
        tiled.size = vec(0, 0)

        result.add(tiled)
        result.add(contents)

        tiled.BUS.hook(GuiComponent.PreDrawEvent::class.java) { event ->
            val box = contents.getLogicalSize()
            tiled.size = vec((box?.width() ?: 0.0) - 1 + 2 * style.BUTTON_BORDER, (box?.height() ?: 0.0) - 1 + 2 * style.BUTTON_BORDER)
        }
    }

}
